# <center> Обучение с учителем. Классификация. Практика

## <center> Прогнозирование оттока клиентов банка

## Постановка задачи

Теперь настало время самостоятельной работы. В рамках самостоятельной работы вам предстоит решить задачу прогнозирования оттока клиентов банка. 

> **Примечание.** Вы уже знакомились с этой задачей в модуле по визуализации («PYTHON-13. Визуализация данных») и провели небольшое исследование данных. Вы можете использовать результаты своего исследования для построения модели машинного обучения.

Напомним контекст задачи:

> Некоторый банк обратился к вам за помощью: он хочет разработать кампанию лояльности по удержанию клиентов. Для этого он хочет прогнозировать вероятности оттока клиентов и определять, уйдёт ли клиент в ближайшее время. 

Итак, ранее вы провели небольшой разведывательный анализ, познакомились с данными и выявили, чем ушедшие клиенты отличаются от лояльных и как между собой связаны различные признаки, определяющие клиентов. Вы отлично справились с первой задачей, и теперь представители банка предлагают вам построить саму модель машинного обучения, которая будет прогнозировать отток клиентов. 

**Ваша задача** - построить классификатор, который позволит своевременно определять уходящих клиентов банка, оценить качество построенных моделей и проинтерпретировать результаты.

Разделим эту задачу на две части:
* В первой части мы подготовим данные для моделирования и построим простейшую модель логистической регрессии, оценим её качество и проанализируем результаты её работы.

* Во второй части мы займёмся построением моделей дерева решений и случайного леса, сравним их результаты с результатами, полученными на первом этапе, и сделаем финальный вывод.


Для начала импортируем библиотеки, которые нам понадобятся:

In [20]:
import numpy as np #для матричных вычислений
import pandas as pd #для анализа и предобработки данных
import matplotlib.pyplot as plt #для визуализации
import seaborn as sns #для визуализации

from sklearn import linear_model #линейные модели
from sklearn import metrics #метрики
from sklearn import model_selection #сплитование выборки
from sklearn import preprocessing #предобработка данных
%matplotlib inline

Итак, прочитаем нашу таблицу:

In [21]:
churn_data = pd.read_csv('data/churn.csv')
churn_data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

## Практика: логистическая регрессия

В первой части вам предстоит выполнить задания 5.0–5.6. Максимальное количество баллов, которое можно получить, — 9:

### Задание 5.0 (не оценивается)

Для полноты задачи проведите небольшой разведывательный анализ данных. Постройте графики распределения целевого признака, диаграммы и графики, показывающие взаимосвязь между исходными признаками и целевым признаком, а также другие диаграммы, иллюстрирующие предоставленные данные. 

**Совет:** воспользуйтесь кодом, который вы писали, выполняя задания в модуле «PYTHON-13. Визуализация данных».

In [22]:
# Ваш код здесь
import plotly.graph_objects as go
from plotly.subplots import make_subplots
# Распределение целевого признака
figure_departed_loyal = go.Figure(data=[go.Pie(labels=['Лояльные', 'Ушедшие'],
                                               values=churn_data['Exited'].value_counts(),
                                               title='Соотношение клиентов')])
# Стиль отображения
figure_departed_loyal.update_traces(textinfo='label+percent', # Содержимое подписи
                                    textfont_size=14, # Размер шрифта подписи
                                    pull=[0.1, 0], # Сдвиг сектора
                                    marker=dict(colors=["#10c938", "#c91010"]), # Цвета секторов
                                    showlegend=False) # Скрыть легенду
# Размер полотна
figure_departed_loyal.update_layout(autosize = False, width = 600, height = 600)
figure_departed_loyal.show()



In [23]:
# Разделение признаков по типам
cat_features = {'Geography': 'Страна проживания клиента',
                'Gender': 'Пол клиента', 
                'Tenure': 'Срок пользования услугами банка',
                'NumOfProducts': 'Количество приобретенных услуг',
                'HasCrCard': 'Наличие кредитной карты',
                'IsActiveMember': 'Активность клиента'}
num_features = {'CreditScore': 'Кредитный рейтинг', 
                'Age': 'Возраст',
                'Balance': 'Суммарный баланс счетов',
                'EstimatedSalary': 'Предполагаемая заработная плата клиента'}

# Цикл построения графиков зависимости для всех признаков из списка категориальных признаков
for cat in cat_features:
    fig = go.Figure()
    # Построение баров
    fig.add_trace(go.Bar(x=churn_data[churn_data['Exited'] == True][cat].value_counts().index,
                         y=churn_data[churn_data['Exited'] == True][cat].value_counts(),
                         marker_color = '#c91010', # Цвет маркера
                         name='Ушедшие')) # Название набора данных
    fig.add_trace(go.Bar(x=churn_data[churn_data['Exited'] == False][cat].value_counts().index,
                         y=churn_data[churn_data['Exited'] == False][cat].value_counts(),
                         marker_color = '#10c938', # Цвет маркера
                         name='Лояльные')) # Название набора данных
    fig.update_layout(autosize = False, width = 800, height = 430, # Размер полотна
                      bargap=0.2) # Расстояние между столбцами
    # Заголовк и подписи осей
    fig.update_layout(title=f'Зависимость оттока клиентов от признака \n"{cat_features[cat]}"',
                      title_x=0.5, # Выравнивание заголовка посередине
                      xaxis_title=cat_features[cat], # Подпись Х
                      yaxis_title='Количество клиентов') # Подпись Y
    fig.show()

# Цикл построения графиков зависимости для всех признаков из списка числовых признаков
for cat in num_features:
    fig = go.Figure()
    fig.add_trace(go.Histogram(x=churn_data[churn_data['Exited'] == True][cat],
                               marker_color = '#c91010', # Цвет маркера
                               nbinsx=100, # Количество столбцов
                               name='Ушедшие'))
    fig.add_trace(go.Histogram(x=churn_data[churn_data['Exited'] == False][cat],
                               marker_color = '#10c938', # Цвет маркера
                               nbinsx=100, # Количество столбцов
                               name='Лояльные'))
    # Размер полотна
    fig.update_layout(autosize = False, width = 1000, height = 620)
    # Заголовк и подписи осей
    fig.update_layout(title=f'Зависимость оттока клиентов от признака \n"{num_features[cat]}"',
                      title_x=0.5, # Выравнивание заголовка посередине
                      xaxis_title=num_features[cat], # Подпись Х
                      yaxis_title='Количество клиентов') # Подпись Y
    fig.show()

В модуле по визуализации мы выяснили, что отток в Германии, Франции и Испании имеет различные показатели. Исследователи, которые уже работали с этим датасетом, пришли к выводу, что наилучшее решение при работе с этими данными — строить модели индивидуально для каждой страны. Мы будем прогнозировать отток в немецком филиале банка. Давайте выделим только данные, относящиеся к Германии:

In [24]:
churn_data = churn_data[churn_data['Geography'] == 'Germany']
churn_data.shape

(2509, 14)

Сразу избавимся от заведомо неинформативных признаков: страна, номер строки, идентификатор пользователя и имя клиента.

In [25]:
churn_data = churn_data.drop(['Geography', 'RowNumber', 'CustomerId', 'Surname'], axis=1)
churn_data.head()

CreditScore  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
7           376  Female   29       4  115046.74              4          1   
15          616    Male   45       3  143129.41              2          0   
16          653    Male   58       1  132602.88              1          1   
26          756    Male   36       2  136815.64              1          1   
28          574  Female   43       3  141349.43              1          1   

    IsActiveMember  EstimatedSalary  Exited  
7                0        119346.88       1  
15               1         64327.26       0  
16               0          5097.67       1  
26               1        170041.95       0  
28               1        100187.43       0

Проверим, что в наших данных нет пропусков:

In [26]:
churn_data.isnull().sum()

CreditScore        0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

Проверим, что в наших данных нет дубликатов:

In [27]:
churn_data[churn_data.duplicated()].sum()

CreditScore          0
Gender               0
Age                  0
Tenure               0
Balance            0.0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
EstimatedSalary    0.0
Exited               0
dtype: object

### Задание 5.1 (1 балл)

Давайте сделаем предобработку данных. 

Для начала расширим датасет с помощью методов Feature Engineering. Создайте следующие признаки:
* `BalanceSalaryRatio` — отношение баланса на счетах к заработной плате клиента;
* `TenureByAge` — отношение времени пользования услугами банка к возрасту клиента;
* `CreditScoreGivenAge` — отношение кредитного рейтинга к возрасту клиента.

Затем перекодируйте оставшиеся текстовые категориальные признаки в числовые бинарные столбцы.

Выведите основные статистические характеристики полученных признаков.

**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**1 балл**): 
    * приведён код для создания указанных в задании признаков;
    * приведён код для кодирования категориальных признаков в числовые бинарные столбцы;
    * приведён код для расчёта основных статистических характеристик полученных признаков.

- Задание выполнено неверно, не учтено одно или несколько условий (**0 баллов**).

In [28]:
# Генерация признаков
churn_data['BalanceSalaryRatio'] = churn_data['Balance'] / churn_data['EstimatedSalary']
churn_data['TenureByAge'] = churn_data['Tenure'] / churn_data['Age']
churn_data['CreditScoreGivenAge'] = churn_data['CreditScore'] / churn_data['Age']

# Кодирование категориальных признаков
churn_data = pd.get_dummies(churn_data, drop_first=True)
display(churn_data.head())
display(churn_data.describe())

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
7           376   29       4  115046.74              4          1   
15          616   45       3  143129.41              2          0   
16          653   58       1  132602.88              1          1   
26          756   36       2  136815.64              1          1   
28          574   43       3  141349.43              1          1   

    IsActiveMember  EstimatedSalary  Exited  BalanceSalaryRatio  TenureByAge  \
7                0        119346.88       1            0.963969     0.137931   
15               1         64327.26       0            2.225020     0.066667   
16               0          5097.67       1           26.012449     0.017241   
26               1        170041.95       0            0.804599     0.055556   
28               1        100187.43       0            1.410850     0.069767   

    CreditScoreGivenAge  Gender_Male  
7             12.965517        False  
15            13.688889         True  
16            11.258621         True  
26            21.000000         True  
28            13.348837        False

CreditScore          Age       Tenure        Balance  NumOfProducts  \
count  2509.000000  2509.000000  2509.000000    2509.000000    2509.000000   
mean    651.453567    39.771622     5.009964  119730.116134       1.519729   
std      98.168937    10.519143     2.935154   27022.006157       0.619420   
min     350.000000    18.000000     0.000000   27288.430000       1.000000   
25%     584.000000    32.000000     2.000000  102800.720000       1.000000   
50%     651.000000    38.000000     5.000000  119703.100000       1.000000   
75%     722.000000    45.000000     8.000000  137560.380000       2.000000   
max     850.000000    84.000000    10.000000  214346.960000       4.000000   

        HasCrCard  IsActiveMember  EstimatedSalary       Exited  \
count  2509.00000     2509.000000      2509.000000  2509.000000   
mean      0.71383        0.497409    101113.435102     0.324432   
std       0.45206        0.500093     58263.011501     0.468256   
min       0.00000        0.000000        11.580000     0.000000   
25%       0.00000        0.000000     51016.020000     0.000000   
50%       1.00000        0.000000    102397.220000     0.000000   
75%       1.00000        1.000000    151083.800000     1.000000   
max       1.00000        1.000000    199970.740000     1.000000   

       BalanceSalaryRatio  TenureByAge  CreditScoreGivenAge  
count         2509.000000  2509.000000          2509.000000  
mean             9.024870     0.134303            17.526487  
std            214.286020     0.087680             5.349558  
min              0.192582     0.000000             6.112676  
25%              0.783284     0.060606            13.686275  
50%              1.197220     0.125000            16.857143  
75%              2.400083     0.200000            20.852941  
max          10614.655440     0.500000            46.388889

Разделим исходный набор данных на матрицу наблюдений `X` (фичи) и столбец ответов `y` (таргет). 

In [29]:
X = churn_data.drop("Exited", axis=1)
y = churn_data["Exited"]

Прежде чем перейти к этапу разделения выборки на тренировочную и тестовую, давайте посмотрим на сбалансированность данных. Для этого воспользуемся методом `value_counts()` и выведем количество наблюдений каждого из классов в долевом соотношении от общего количества наблюдений:

In [30]:
y.value_counts(normalize=True)

Exited
0    0.675568
1    0.324432
Name: proportion, dtype: float64

Очевидно, что классы не сбалансированы. Ушедших пользователей 32 %, в то время как лояльных — 68 %.


**К чему это может привести?**

При разделении набора данных на тренировочный и тестовый может возникнуть такая ситуация, что в тренировочную выборку попадут только лояльные клиенты или их большая часть. Тогда модель может не научиться определять вероятность оттока и качество на тестовой выборке будет неудовлетворительным.

Или противоположная ситуация: в тестовую выборку попадут только лояльные клиенты. Тогда модель будет показывать отличное качество на тестовой выборке, но в реальных условиях она может оказаться бесполезной.

То есть нам нужно сделать так, чтобы в тренировочной и тестовой выборке было одинаковое соотношение лояльных и ушедших клиентов. Такое разбиение выборки называется **стратифицированным**. 

Функция `train_test_split()` из библиотеки `sklearn` умеет производить такое разбиение. Для этого в ней предусмотрен параметр `stratify`, который нужно выставить в значение признака, который должен иметь одинаковые соотношения в тренировочной и тестовой выборке. У нас этот признак - это целевой признак y. 

Выполним разбиение, установив значение параметра `random_state` на 0:


In [31]:
X_train, X_test, y_train, y_test = model_selection.train_test_split(
    X, y, 
    stratify=y, 
    random_state=0
)
print('Train shape: {}'.format(X_train.shape))
print('Test shape: {}'.format(X_test.shape))

Train shape: (1881, 12)
Test shape: (628, 12)


Проверим, что число лояльных и ушедших клиентов примерно одинаково в каждой из выборок:

In [32]:
print('Train :')
display(y_train.value_counts(normalize=True))
print('\n')
print('Test :', )
display(y_test.value_counts(normalize=True))

Train :


Exited
0    0.675704
1    0.324296
Name: proportion, dtype: float64



Test :


Exited
0    0.675159
1    0.324841
Name: proportion, dtype: float64

Соотношения одинаковы, а значит, мы можем перейти к следующему шагу.

### Задание 5.2 (1 балл)

Позаботьтесь о масштабировании признаков. Проведите нормализацию/стандартизацию матрицы наблюдений `X` с помощью `MinMaxScaler`/`RobustScaler` или `StandardScaler`. Выберите метод самостоятельно.

Обучите один из предложенных скейлеров на тренировочной выборке (вычислите параметры трансформации) и примените трансформацию данных к тренировочной и тестовой выборкам.

**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**1 балл**): 
    * приведён код для создания одного из вариантов скейлеров для масштабирования данных;
    * приведён код для расчёта параметров трансформации (обучение скейлера) на тренировочной выборке;
    * приведён код для трансформации тренировочных и тестовых данных (признаков) в соответствии с вычисленными параметрами.

- Задание выполнено неверно, не учтено одно или несколько условий (**0 баллов**).

In [33]:
# Отсутствуют сведения об очистке данных от выбросов, выбираем 
# слабочувствительный к выбросам метод нормализации RobustScaler
scaler = preprocessing.RobustScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Метод RobustScaler не масштабирует данных, проводим стандартизацию
std_scaler = preprocessing.StandardScaler()
std_scaler.fit(X_train_scaled)
X_train_sc_std = std_scaler.transform(X_train_scaled)
X_test_sc_std = std_scaler.transform(X_test_scaled)

Теперь переходим к моделированию.

### Задание 5.3 (1 балл)

Для начала давайте заранее определим метрику, на которую будем ориентироваться. 

По условию задачи, наша модель должна своевременно предсказывать отток клиентов. Известно, что для сохранения клиентов банк будет использовать ресурсы колл-центра и клиентам, у которых вероятность оттока наиболее высока, будут сделаны специальные предложения. Ресурсы, выделяемые на сохранение клиентов, ограничены, поэтому нам важно минимизировать количество звонков. Однако в то же время необходимо сохранить как можно больше клиентов.

Таким образом, нам бы хотелось минимизировать как ложноположительные и ложноотрицательные срабатывания модели.

Какую метрику в данной задаче вы будете пытаться улучшить? Обоснуйте свой выбор. 


> Укажите название метрики и обоснования своего выбора здесь

**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**1 балл**): 
    * выбрана верная метрика;
    * приведено развернутое обоснование выбора метрики, и приведенные рассуждения являются верными;

- Задание выполнено неверно, не учтено одно или несколько условий (**0 баллов**).

> Ранее было подчеркнуто, что в данной выборке классы не сбалансированы, таким образом, использование метрики accuracy приведет к получению недостоверной оценки.  
Так как по условиям задачи требуется минимизировать как ложноположительные, так и ложноотрицательные срабатывания модели, то необходимо сбалансировать метрики точности (precision) и полноты (recall) предсказаний, что достигается подбором оптимального порога вероятности при рассмотрении значений метрик precision, recall и f-меры. При этом в условии отсутствия данных о соотношении экономических затрат на удержание клиента и экономического эффекта от сохранения его лояльности значение оптимального порога вероятности можно принять такое, при котором precision, recall и f-мера равны/максимально приближены.

> Ранее было подчеркнуто, что в данной выборке классы не сбалансированы, таким образом, использование метрики accuracy приведет к получению недостоверной оценки модели.  
Так как по условиям задачи требуется минимизировать как ложноположительные, так и ложноотрицательные срабатывания, то необходимо использовать сбалансированную оценку метрик точности (precision) и полноты (recall) предсказаний, такую оценку дает **F-мера**.

### Задание 5.4 (2 балла)

Обучите модель логистической регрессии (`Logistic Regression`) на тренировочных данных. Позаботьтесь о воспроизводимости результата, установив конкретное значение параметра `random_state` (например, число 42).

Сделайте предсказание меток классов для тренировочной и тестовой выборок. 

Рассчитайте значение целевой метрики на тренировочной и тестовой выборках. 

Проинтерпретируйте полученные результаты, ответив на следующий вопрос:
* Как вы считаете, является ли ваша модель переобученной или недообученной?

> **Примечание.** Будем считать, что модель нас не удовлетворяет (считается недообученной), если значение целевой метрики на тестовой выборке меньше 0.5.

**Дополнительное задание:**

Попробуйте поуправлять следующими параметрами модели:
* `penalty` — тип используемой регуляризации ('l1' — $L_1$-регуляризация, 'l2' — $L_2$-регуляризация);
* `C` — коэффициент, обратный коэффициенту регуляризации (по умолчанию он равен 1).

Понаблюдайте, как ведут себя метрики в зависимости от типа регуляризации и увеличения/уменьшения значения парамера `C`.

**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**2 балла**): 
    * построена модель логистической регрессии, решающая поставленную задачу;
    * с помощью полученной модели сформированы предсказания на тренировочной и тестовой выборках;
    * приведён расчёт выбранной метрики на тренировочных и тестовых данных (метрика должна быть выбрана корректно);
    * приведён небольшой вывод по полученным результатам, в котором содержится ответ на поставленный в задании вопрос.

- Задание выполнено верно, но не приведён вывод по полученным результатам или вывод является неверным (**1 балл**).
- Задание выполнено неверно, не учтено несколько условий (**0 баллов**).

In [34]:
def show_figure_LogisticRegression_var(X_train, y_train, X_test, y_test, max_C):
    # Словарь алгоритмов оптимизации и поддерживаемых им способов регуляризации
    methods = {'lbfgs': ['l2'],
               'liblinear': ['l1', 'l2'],
               'newton-cg': ['l2'],
               'newton-cholesky': ['l2'],
               'sag': ['l2'],
               'saga': ['l1', 'l2']}
    # Вычисление и отображение на графике метрик
    for method in methods:
        fig = go.Figure()
        for regl in methods[method]:
            f1s = {}
            for c in np.arange(0.1, max_C, 0.1, dtype=np.float16):
                log_reg = linear_model.LogisticRegression(solver=method, #алгоритм оптимизации
                                                          penalty=regl,
                                                          C=c,
                                                          random_state=42, #генератор случайных чисел
                                                          max_iter=1000)
                # Обучаем модель, минимизируя logloss
                log_reg.fit(X_train, y_train)
                # Предсказание для тестовой выборки
                y_test_pred = log_reg.predict(X_test)
                f1s[c] = metrics.f1_score(y_test, y_test_pred)
            # Построение графиков
            fig.add_trace(go.Scatter(x=list(f1s.keys()),
                                     y=list(f1s.values()),
                                     showlegend=True,
                                     name=regl))
        fig.update_layout(autosize = False, width = 600, height = 400, title=method, title_x=0.5)
        fig.update_xaxes(title_text='C')
        fig.update_yaxes(title_text='F1-мера')
        fig.show()
        
        
show_figure_LogisticRegression_var(X_train_sc_std, y_train, X_test_sc_std, y_test, 2.5)

In [35]:
# Создание объекта класса логистическая регрессия
log_reg = linear_model.LogisticRegression(
    solver='saga', #алгоритм оптимизации
    penalty='l1',
    random_state=42, #генератор случайных чисел
    max_iter=1000 #количество итераций на сходимость
)
# Обучение модели
log_reg.fit(X_train_sc_std, y_train)

# Предсказание для тренировочной выборки
y_train_pred = log_reg.predict(X_train_sc_std)
print(f'F1 для тренировочной выборки: {metrics.f1_score(y_train, y_train_pred):.3f}')

# Предсказание для тестовой выборки
y_test_pred = log_reg.predict(X_test_sc_std)
print(f'F1 для тестовой выборки: {metrics.f1_score(y_test, y_test_pred):.3f}')

F1 для тренировочной выборки: 0.515
F1 для тестовой выборки: 0.496


>Максимальным значением $F_1$-меры для тестовой выборки является **0.496**.  
Оно может быть получено:
>- при использовании алгоритмов оптимизации *lbfgs*, *newton-cg*, *newton-cholesky* и *sag* при *С* равном *0.5* (алгоритмы поддерживают только $L_2$-регуляризацию);
>- при использовании алгоритма оптимизации *liblinear* и $L_2$-регуляризации при *С* в диапазоне от 1.3 до 2.2;
>- при использовании алгоритма оптимизации *saga* и $L_2$-регуляризации при *С* в диапазоне от 0.5 до 0.6;
>- при использовании алгоритма оптимизации *saga* и $L_1$-регуляризации при *С* в диапазоне от 1 до 1.8.
>
>Так как полученное значение хоть и незначительно, но меньше установленного порога, то считаем текущую модель ***недообученной***.

### Задание 5.5 (2 балла)

Сгенерируйте полиномиальные признаки третьей степени для тренировочной и тестовой выборок. Используйте класс `PolynomialFeatures` из библиотеки `sklearn`. Параметр `include_bias` выставите в значение `False`.

> **Примечание.** Обратите внимание, что для генерации полиномов необходимо использовать масштабированные данные.

Обучите модель логистической регрессии на полиномиальных признаках (`Logistic Regression`). Позаботьтесь о воспроизводимости результат, установив конкретное значение параметра `random_state` (например, число 42).

Предскажите метки классов для тренировочной и тестовой выборок.

Рассчитайте значение целевой метрики на тренировочной и тестовой выборках для модели, обученной на полиномиальных признаках.

Сравните результат с полученным ранее. Удалось ли вам улучшить качество вашей модели?

**Дополнительное задание:**

Попробуйте поуправлять следующими параметрами модели:
* `penalty` — тип используемой регуляризации ('l1' — $L_1$-регуляризация, 'l2' — $L_2$-регуляризация);
* `C` — коэффициент, обратный коэффициенту регуляризации (по умолчанию он равен 1).

Понаблюдайте, как ведут себя метрики в зависимости от типа регуляризации и увеличения/уменьшения значения парамера `C`.

**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**2 балла**): 
    * сгенерированы полиномиальные признаки третьей степени;
    * построена модель логистической регрессии, решающая поставленную задачу на основе сгенерированных признаков;
    * с помощью полученной модели сформированы предсказания на тренировочной и тестовой выборках;
    * приведён расчёт выбранной метрики на тренировочных и тестовых данных (метрика должна быть выбрана корректно);
    * приведён небольшой вывод по полученным результатам, в котором содержится корректный ответ на поставленный в задании вопрос.

- Задание выполнено верно, но не приведён вывод по полученным результатам или вывод является неверным (**1 балл**).
- Задание выполнено неверно, не учтено несколько условий (**0 баллов**).

In [36]:
# Генерация полиномиальных признаков
pf = preprocessing.PolynomialFeatures(degree=3,
                                      include_bias=False)
pf.fit(X_train_sc_std)
X_train_sc_std_pf = pf.transform(X_train_sc_std)
X_test_sc_std_pf = pf.transform(X_test_sc_std)
# Отключение предупреждений
import warnings
warnings.filterwarnings("ignore")
# Построение графиков
show_figure_LogisticRegression_var(X_train_sc_std_pf, y_train, X_test_sc_std_pf, y_test, 1.5)

In [40]:
print(f'Количество признаков: {X_train_sc_std_pf.shape[1]}')
# Создание объекта класса логистическая регрессия
log_reg = linear_model.LogisticRegression(
    solver='liblinear', #алгоритм оптимизации
    penalty='l1',
    C=0.2,
    random_state=42, #генератор случайных чисел
    max_iter=1000 #количество итераций на сходимость
)
# Обучаем модель, минимизируя logloss
log_reg.fit(X_train_sc_std_pf, y_train)

# Предсказание для тренировочной выборки
y_train_pred = log_reg.predict(X_train_sc_std_pf)
print(f'F1 для тренировочной выборки: {metrics.f1_score(y_train, y_train_pred):.3f}')

# Предсказание для тестовой выборки
y_test_pred = log_reg.predict(X_test_sc_std_pf)
print(f'F1 для тестовой выборки: {metrics.f1_score(y_test, y_test_pred):.3f}')

Количество признаков: 454
F1 для тренировочной выборки: 0.741
F1 для тестовой выборки: 0.683


> После генерации полиномиальных признаков третьей степени получено 454 фактора.  
Наилучшее значение $F_1$-меры равно **0.683**, было получено при использовании алгоритма оптимизации *liblinear* и $L_1$-регуляризации при *С* равном 0.2. В ходе тестирования поддерживаемых алгоритмов оптимизации на этапах вычислений с применением алгоритмов *sag* и *saga* были получены сообщения о достижении предельного числа итераций (позже предупреждения были скрыты), что свидетельствует о низкой сходимости данных алгоритмов при таком количестве факторов тренировочной и тестовой выборок.  
Построенная модель логистической регрессии имеет значение $F_1$-меры, удовлетворяющее условиям задания, этот факт позволяет сделать вывод, что генерация полиномиальных признаков помогла существенно улучшить качество модели.

### Задание 5.6. (2 балла)


Попробуйте подобрать значение порога вероятности, при котором наблюдается наибольшее значение целевой метрики на тестовой выборке. В качестве модели используйте логистическую регрессию, обученную на полиномиальных признаках. 

В качестве порога вероятности используйте значения из следующего диапазона:

`thresholds = np.arange(0.1, 1, 0.05)`

Напомним алгоритм, как определить оптимальный порог вероятности:
* Предсказать вероятности оттока для каждого клиента из тестовой выборки.
* В цикле перебрать значения порогов вероятности threshold:
    * Сделать предсказание класса по принципу: если вероятность > threshold, то отнести объект к классу 1, в противном случае — к классу 0.
    * Рассчитать метрику и занести её в заранее созданный список.

Для реализации алгоритма подбора вы можете воспользоваться шаблоном кода, приведённым ниже:

```python
# Делаем предсказание вероятности оттока для каждого клиента из тестовой выборки
y_test_proba_pred = pd.Series(model.predict_proba(X_test_poly)[:, 1])
# Инициализируем список, куда будем сохранять метрики.
scores = []
# Задаём значения порогов вероятности
thresholds = np.arange(0.1, 1, 0.05)
# В цикле перебираем значения порогов вероятности
for threshold in thresholds:
    # Делаем предсказание
    y_test_pred_poly = y_test_proba_pred.apply(lambda x: 1 if x > threshold else 0)
    # Рассчитываем метрику
    score = # код для вычисления метрики
    # Добавляем результат в список
    scores.append(score)
```

Постройте график зависимости целевой метрики от порога вероятности. На основе построенного графика определите порог вероятности, которому соответствует наибольшее значение целевой метрики на тестовой выборке.

Сделайте предсказание классов с подобранным оптимальным порогом вероятности. Рассчитайте значение целевой метрики на тестовой выборке и сравните результат с полученными ранее.


**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**2 балла**): 
    * приведён код для построения графика зависимости значения выбранной метрики от порога вероятности;
    * найден оптимальный порог вероятности, на котором достигается наибольшее значение выбранной метрики на тестовой выборке;
    * на основе найденного порога вероятности построено предсказание модели для тренировочной и тестовой выборок;
    * приведён расчёт выбранной метрики на тренировочных и тестовых данных (метрика должна быть выбрана корректно).

- Задание выполнено верно, но не приведён вывод по полученным результатам или вывод является неверным (**1 балл**).
- Задание выполнено неверно, не учтено несколько условий (**0 баллов**).

In [ ]:
# Получение вероятностей принадлежности к классу
y_test_proba_pred = pd.Series(log_reg.predict_proba(X_test_sc_std_pf)[:, 1])
scores = []
# Генерация набора вероятностных порогов
thresholds = np.arange(0.1, 1, 0.05)
# Вычисление метрик в зависимости от применения пороговых вероятностей
for threshold in thresholds:
    y_test_pred = y_test_proba_pred.apply(lambda x: 1 if x>threshold else 0)
    scores.append(metrics.f1_score(y_test, y_test_pred))

# Построение графика
fig = go.Figure(data=go.Scatter(x=thresholds,
                                y=scores))
fig.update_layout(autosize = False, width = 600, height = 400, 
                  title='Зависимость F1-меры от пороговой вероятности', title_x=0.5)
fig.update_xaxes(title_text='Пороговая вероятность')
fig.update_yaxes(title_text='F1-мера')
fig.show()

In [48]:
# Предсказание для тренировочной выборки
y_train_proba_pred = pd.Series(log_reg.predict_proba(X_train_sc_std_pf)[:, 1])
y_train_pred = y_train_proba_pred.apply(lambda x: 1 if x > 0.4 else 0)
print(f'F1 для тренировочной выборки: {metrics.f1_score(y_train, y_train_pred):.3f}')

# Предсказание для тестовой выборки
y_test_proba_pred = pd.Series(log_reg.predict_proba(X_test_sc_std_pf)[:, 1])
y_test_pred = y_test_proba_pred.apply(lambda x: 1 if x > 0.4 else 0)
print(f'F1 для тестовой выборки: {metrics.f1_score(y_test, y_test_pred):.3f}')

F1 для тренировочной выборки: 0.759
F1 для тестовой выборки: 0.688


>Оптимальный порог вероятности составил **0.4**, при данном пороге достигается наибольшее значение $F_1$-меры, равное **0.688**, что несколько улучшило качество получаемых спрогнозированных данных.

## Практика: деревья решений и случайный лес

Мы продолжаем решать задачу прогнозирования оттока клиентов банка.

Ранее мы построили модель логистической регрессии, затем добавили в модель полиномиальные признаки и подобрали оптимальный порог вероятности для логистической регрессии.

> Теперь попробуем решить задачу с помощью деревьев решений и лесов.


> **Примечание №1.** Для выполнения дальнейших заданий необходимо совершить уже знакомые шаги предобработки с данными об оттоке клиентов банка, а именно:
* выделить данные, относящиеся к отделениям банка в Германии;
* избавиться от заведомо неинформативных признаков;
* расширить датасет путём добавления новых признаков;
* разделить выборку на тренировочную и тестовую со стратификацией.

> **Примечание №2.** Деревья решений — алгоритм, способный уловить нелинейные зависимости. Генерация полиномиальных признаков практически не имеет значения.
Также деревья решений не чувствительны к масштабу признаков. Масштабирование исходных данных не играет роли при построении деревьев решения.
Поэтому для решения задачи используйте обучающую (`X_train`, `y_train`) и тестовую (`X_test`, `y_test`) выборки без полиномиальных столбцов.


In [ ]:
X_train, X_test, y_train, y_test = model_selection.train_test_split(
    X, y, 
    stratify=y, 
    random_state=0
)
print('Train shape: {}'.format(X_train.shape))
print('Test shape: {}'.format(X_test.shape))

Импортируем необходимые для выполнения практики модули:

In [ ]:
from sklearn import tree
from sklearn import ensemble

Приступим! Выполните задания 8.1–8.4.

### Задание 8.1. (2 балла)

Обучите на тренировочной выборке дерево решений из библиотеки `sklearn` (`DecisionTreeClassifier`) с неограниченной максимальной глубиной дерева. В качестве критерия информативности используйте энтропию Шеннона, остальные параметры оставьте по умолчанию. 

Также позаботьтесь о воспроизводимости результатов, установив конкретное  значение `random_state` (например, число 42). 

Сделайте предсказания для тренировочной и тестовой выборок. Рассчитайте значение целевой метрики на каждой из выборок.

Сделайте вывод, ответив на вопрос:
* Как вы считаете, является полученная модель переобученной или недообученной? Попробуйте дать обоснование, почему так происходит.


**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**2 балла**): 
    * построена модель дерева решений с заданными в условии задачи параметрами;
    * с помощью полученной модели сформированы предсказания для тренировочной и тестовой выборок;
    * приведён расчёт выбранной метрики на тренировочных и тестовых данных (метрика должна быть выбрана корректно);
    * приведён небольшой вывод по полученным результатам, в котором содержится корректный ответ на поставленный в задании вопрос, а также обоснование полученного результата.

- Задание выполнено верно, но не приведено обоснование полученного результата или обоснование является некорректным (**1 балл**).
- Задание выполнено неверно, не учтено несколько условий (**0 баллов**).

In [ ]:
# Ваш код здесь

> Ваши выводы и рассуждения — *здесь*

### Задание 8.2. (1 балл)

Давайте «подстрижём» наше дерево. 

Обучите на тренировочной выборке дерево решений из библиотеки `sklearn` (`DecisionTreeClassifier`) с максимальной глубиной 8. В качестве критерия информативности используйте энтропию Шеннона, минимальное число объектов в листе (количество объектов, при которых дерево прекратит делиться и образуется лист) — 10.

Также позаботьтесь о воспроизводимости результатов, установив конкретное значение `random_state` (например, число 42).

Сделайте предсказания для тренировочной и тестовой выборок. Рассчитайте значение целевой метрики на каждой из выборок.

Сравните результат с полученными ранее. Сделайте вывод.


**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**1 балла**): 
    * построена модель дерева решений с заданными в условии задачи параметрами;
    * с помощью полученной модели сформированы предсказания для тренировочной и тестовой выборок;
    * приведён расчёт выбранной метрики на тренировочных и тестовых данных (метрика должна быть выбрана корректно);
    * приведён небольшой вывод по полученным результатам, в котором содержится корректный ответ на поставленный в задании вопрос.

- Задание выполнено неверно, не учтено одно или несколько условий (**0 баллов**).

In [ ]:
# Ваш код здесь

> Ваши выводы и рассуждения — *здесь*

### Задание 8.3. (1 балл)

Давайте воспользуемся случайным лесом. 

Постройте случайный лес (`RandomForestClassifier`) из 500 деревьев с максимальной глубиной 8. Критерий информативности — энтропия Шеннона, минимальное число объектов в листе (количество объектов, при которых дерево прекратит делиться и образуется лист) — 10.

Также позаботьтесь о воспроизводимости результатов, установив конкретное значение значение `random_state` (например, число 42).

Сделайте предсказания для тренировочной и тестовой выборок. Рассчитайте значение целевой метрики на каждой из выборок.

Сравните качество одного дерева решений с качеством случайного леса на тестовой выборке. Смогли ли вы улучшить метрику?


**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**1 балл**): 
    * построена модель дерева решений с заданными в условии задачи параметрами;
    * с помощью полученной модели сформированы предсказания для тренировочной и тестовой выборок;
    * приведён расчёт выбранной метрики на тренировочных и тестовых данных (метрика должна быть выбрана корректно);
    * приведён небольшой вывод по полученным результатам, в котором содержится корректный ответ на поставленный в задании вопрос.

- Задание выполнено неверно, не учтено одно или несколько условий (**0 баллов**).

In [ ]:
# Ваш код здесь

> Ваши выводы и рассуждения — *здесь*

### Задание 8.4. (2 балла)

Мы на финишной прямой. 


Попробуйте подобрать значение порога вероятности, при котором наблюдается наибольшее значение целевой метрики на тестовой выборке. **Используйте ту модель (из рассмотренных ранее), которая показала наилучшее качество на тестовой выборке.**


В качестве порога вероятности используйте значения из следующего диапазона:

`thresholds = np.arange(0.1, 1, 0.05)`

Напомним алгоритм, как определить оптимальный порог вероятности:
* Предсказать вероятности оттока для каждого клиента из тестовой выборки.
* В цикле перебрать значения порогов вероятности threshold:
    * Сделать предсказание класса по принципу: если вероятность > threshold, то отнести объект к классу 1, в противном случае — к классу 0.
    * Рассчитать метрику и занести её в заранее созданный список. 

Для реализации алгоритма подбора вы можете воспользоваться шаблоном кода, приведённым ниже:

```python
# Делаем предсказание вероятности оттока для каждого клиента из тестовой выборки
y_test_proba_pred = pd.Series(model.predict_proba(X_test_poly)[:, 1])
# Инициализируем список, куда будем сохранять метрики.
scores = []
# Задаём значения порогов вероятности
thresholds = np.arange(0.1, 1, 0.05)
# В цикле перебираем значения порогов вероятности
for threshold in thresholds:
    # Делаем предсказание
    y_test_pred_poly = y_test_proba_pred.apply(lambda x: 1 if x > threshold else 0)
    # Рассчитываем метрику
    score = # код для вычисления метрики
    # Добавляем результат в список
    scores.append(score)
```

Постройте график зависимости целевой метрики от порога вероятности. На основе построенного графика определите порог вероятности, которому соответствует наибольшее значение целевой метрики на тестовой выборке.

Сделайте предсказание классов с подобранным оптимальным порогом вероятности. Рассчитайте значение целевой метрики на тестовой выборке и сравните полученный результат с полученными ранее. 

Напишите небольшой вывод о проделанной работе, указав в нём, какая из рассмотренных нами моделей показала наилучший результат.

> **Примечание.** Помните о том, что в реальных задачах (а мы решали задачу на реальных данных) мы практически никогда не сможем получить значения метрик, близкие к 1. Если в задаче прогнозирования оттока ваша метрика превышает хотя бы значения 0.6-0.7, вашу модель уже можно считать хорошей и можно говорить о том, что она успешно справляется с задачей.

In [ ]:
# Ваш код здесь

> Ваши выводы и рассуждения — *здесь*

### Задание 8.5. (1 балл)

Наконец, финальная задача.

У нашего банка есть некоторый клиент, назовём его Василием. В базе данных банка хранятся следующие данные о Василии:

```python
{
    'CreditScore': [601.0],
    'Gender': ['Male'],
    'Age': [42.0],
    'Tenure': [1.0],
    'Balance': [98495.72],
    'NumOfProducts': [1.0],
    'HasCrCard': [1.0],
    'IsActiveMember': [0.0],
    'EstimatedSalary': [40014.76]
}
```

С помощью наилучшей модели спрогнозируйте вероятность того, что Василий откажется от услуг банка.

> **Примечание.** Учтите, что для предсказания необходимо преобразовать данные к тому формату, на котором обучалась модель, то есть:
* добавить признаки дополнительные признаки; 
* произвести кодирование категориальных признаков;
* произвести масштабирование признаков (если ваша модель обучалась на масштабированных данных) с помощью того же скейлера, который вы использовали при обучении модели;
* сгенерировать полиномиальные признаки.

**Критерии оценивания:**
- Задание выполнено верно, учтены все условия (**1 балл**): 
    * сформирован DataFrame на основе предоставленного словаря;
    * произведены преобразования данных, необходимые для работы модели;
    * предсказана вероятность оттока клиента (вероятность принадлежности к классу 1).

- Задание выполнено неверно, не учтено одно или несколько условий (**0 баллов**).

In [ ]:
# Ваш код здесь